Mathlib/GroupTheory/SpecificGroups/KleinFour.lean

```diff
@@ -39,74 +40,195 @@ produces the third one.
-  toFun := fun
-    | (0, 0) => 1
-    | (0, 1) => sr 1
-    | (1, 0) => r 1
-    | (1, 1) => sr 1 * r 1
-  invFun := fun
-    | 1 => (0, 0)
-    | sr 1 => (0, 1)
-    | r 1 => (1, 0)
-    | sr 1 * r 1 => (1, 1)
-  left_inv := by decide
-  right_inv := by decide
-  map_mul' := by simp (config := {decide := true})
-
-end KleinFour
+section ExponentTwo
+
+variable {G : Type*} [Group G]
+
+/-- In a group of exponent two, every element is its own inverse. -/
+@[to_additive]
+lemma inv_eq_self_of_exponent_two (hG : Monoid.exponent G = 2) (x : G) :
+    x⁻¹ = x :=
+  inv_eq_of_mul_eq_one_left <| pow_two (a := x) ▸ hG ▸ Monoid.pow_exponent_eq_one x
+
+/-- If an element in a group has order two, then it is its own inverse. -/
+@[to_additive]
+lemma inv_eq_self_of_orderOf_eq_two {x : G} (hx : orderOf x = 2) :
+    x⁻¹ = x :=
+  inv_eq_of_mul_eq_one_left <| pow_two (a := x) ▸ hx ▸ pow_orderOf_eq_one x
+
+@[to_additive]
+lemma orderOf_eq_two_iff (hG : Monoid.exponent G = 2) {x : G} :
+    orderOf x = 2 ↔ x ≠ 1 :=
+  ⟨by rintro hx rfl; norm_num at hx, orderOf_eq_prime (hG ▸ Monoid.pow_exponent_eq_one x)⟩
+
+/-- In a group of exponent two, all elements commute. -/
+@[to_additive]
+lemma mul_comm_of_exponent_two (hG : Monoid.exponent G = 2) (x y : G) :
+    x * y = y * x := by
+  simpa only [inv_eq_self_of_exponent_two hG] using mul_inv_rev x y
+
+/-- Any group of exponent two is abelian. -/
+@[to_additive (attr := reducible) "Any additive group of exponent two is abelian."]
+def instCommGroupOfExponentTwo (hG : Monoid.exponent G = 2) : CommGroup G where
+  mul_comm := mul_comm_of_exponent_two hG
+
+@[to_additive]
+lemma mul_not_mem_of_orderOf_eq_two {G : Type*} [Group G] {x y : G} (hx : orderOf x = 2)
+    (hy : orderOf y = 2) (hxy : x ≠ y) : x * y ∉ ({x, y, 1} : Set G) := by
+  simp only [Set.mem_singleton_iff, Set.mem_insert_iff, mul_right_eq_self, mul_left_eq_self, not_or]
+  refine ⟨?_, ?_, fun h' ↦ hxy <| by
+    simpa only [mul_eq_one_iff_eq_inv, inv_eq_self_of_orderOf_eq_two hx,
+      inv_eq_self_of_orderOf_eq_two hy] using h'⟩
+  all_goals rw [← orderOf_eq_one_iff]; norm_num [hy, hx]
+
+@[to_additive]
+lemma mul_not_mem_of_exponent_two {G : Type*} [Group G] (h : Monoid.exponent G = 2) {x y : G}
+    (hx : x ≠ 1) (hy : y ≠ 1) (hxy : x ≠ y) : x * y ∉ ({x, y, 1} : Set G) :=
+  mul_not_mem_of_orderOf_eq_two (orderOf_eq_prime (h ▸ Monoid.pow_exponent_eq_one x) hx)
+    (orderOf_eq_prime (h ▸ Monoid.pow_exponent_eq_one y) hy) hxy
+
+end ExponentTwo
+
+/-! # Klein four-groups as a mixin class -/
+
+/-- An (additive) Klein four-group is an (additive) group of cardinality four and exponent two. -/
+class IsAddKleinFour (G : Type*) [AddGroup G] : Prop where
+  card_four : Nat.card G = 4
+  exponent_two : AddMonoid.exponent G = 2
+
+/-- A Klein four-group is a group of cardinality four and exponent two. -/
+@[to_additive existing IsAddKleinFour]
+class IsKleinFour (G : Type*) [Group G] : Prop where
```

Comment:
There is a zulip [discussion](https://leanprover.zulipchat.com/#narrow/stream/345428-mathlib-reviewers/topic/Klein.204) on making this a class or not. (I said to make it just a predicate, but I don't have a very strong opinion).